Loop/Managers/DeviceDataManager.swift

```diff
@@ -1291,100 +1291,110 @@ extension Notification.Name {
+            self.enactBolus(units: bolusAmount, activationType: .manualNoRecommendation) { error in
+                if let error = error {
+                    NotificationManager.sendRemoteBolusFailureNotification(for: error, amount: bolusAmount)
+                } else {
+                    NotificationManager.sendRemoteBolusNotification(amount: bolusAmount)
+                }
+            }
+        case .carbsEntry(let candidateCarbEntry):
+            log.default("Adding carbs entry.")
+            
+            let candidateCarbsInGrams = candidateCarbEntry.quantity.doubleValue(for: .gram())
+            
+            //Remote carb entry requires validation from its remote source
+            guard remoteDataServicesManager.validatePushNotificationSource(notification) else {
+                NotificationManager.sendRemoteCarbEntryFailureNotification(for: RemoteCommandError.invalidOTP, amountInGrams: candidateCarbsInGrams)
+                log.info("Could not validate notification: %{public}@", String(describing: notification))
+                return
+            }
+            
+            guard candidateCarbsInGrams > 0.0 else {
+                NotificationManager.sendRemoteCarbEntryFailureNotification(for: RemoteCommandError.invalidCarbs, amountInGrams: candidateCarbsInGrams)
+                log.default("Invalid carb entry amount. Aborting...")
+                return
+            }
+            
+            guard candidateCarbsInGrams <= LoopConstants.maxCarbEntryQuantity.doubleValue(for: .gram()) else {
+                NotificationManager.sendRemoteCarbEntryFailureNotification(for: RemoteCommandError.exceedsMaxCarbs, amountInGrams: candidateCarbsInGrams)
+                log.default("Carbs higher than maximum. Aborting...")
+                return
+            }
+            
+            carbStore.addCarbEntry(candidateCarbEntry) { carbEntryAddResult in
+                switch carbEntryAddResult {
+                case .success(let completedCarbEntry):
+                    NotificationManager.sendRemoteCarbEntryNotification(amountInGrams: completedCarbEntry.quantity.doubleValue(for: .gram()))
+                case .failure(let error):
+                    NotificationManager.sendRemoteCarbEntryFailureNotification(for: error, amountInGrams: candidateCarbsInGrams)
                 }
-                // Wait up to 25 seconds for uploads triggered by these commands to finish
-                let _ = remoteDataServicesManager.waitForUploadsToFinish(timeout: .now() + TimeInterval(25))
-            } else {
-                log.error("Unhandled remote notification: %{public}@", String(describing: notification))
             }
         }
-        log.info("Finished handling remote notification")
+        // Wait up to 25 seconds for uploads triggered by these commands to finish
+        let _ = remoteDataServicesManager.waitForUploadsToFinish(timeout: .now() + TimeInterval(25))
```

Comment:
Unhandled notifications errors will now be thrown by the RemoteCommand initializer.